packages/core/core/src/requests/AssetRequestRunner.js

```diff
@@ -0,0 +1,155 @@
+    optionsRef: number,
+    configRef: number,
+    workerFarm: WorkerFarm,
+    assetGraph: AssetGraph,
+  |}) {
+    super(opts);
+    this.type = 'asset_request';
+    this.options = opts.options;
+    this.optionsRef = opts.optionsRef;
+    this.configRef = opts.configRef;
+    this.runTransform = opts.workerFarm.createHandle('runTransform');
+    this.assetGraph = opts.assetGraph;
+  }
+
+  async run(request: AssetRequestDesc, api: RequestRunnerAPI) {
+    api.invalidateOnFileUpdate(
+      await this.options.inputFS.realpath(request.filePath),
+    );
+    let start = Date.now();
+    let {assets, configRequests} = await this.runTransform({
+      request: request,
+      optionsRef: this.optionsRef,
+      configRef: this.configRef,
+    });
+
+    let time = Date.now() - start;
+    for (let asset of assets) {
+      asset.stats.time = time;
+    }
+    return {assets, configRequests};
+  }
+
+  onComplete(
+    request: AssetRequestDesc,
+    result: AssetRequestResult,
+    api: RequestRunnerAPI,
+  ) {
+    this.assetGraph.resolveAssetGroup(request, result.assets);
+
+    let {assets, configRequests} = result;
+
+    for (let asset of assets) {
+      for (let filePath of asset.includedFiles.keys()) {
+        api.invalidateOnFileUpdate(filePath);
+        api.invalidateOnFileDelete(filePath);
+      }
+    }
+
+    // TODO: this should no longer be needed once we have ConfigRequestRunner
+    let graph = this.tracker.graph;
+    let subrequestNodes = [];
+    // Add config requests
+    for (let {request, result} of configRequests) {
+      let id = generateRequestId('config_request', request);
+      let shouldSetupInvalidations =
+        graph.invalidNodeIds.has(id) || !graph.hasNode(id);
+      let subrequestNode = nullthrows(
+        graph.addRequest({
+          id,
+          type: 'config_request',
+          request,
+          result,
+        }),
+      );
+      invariant(subrequestNode.type === 'request');
+
+      if (shouldSetupInvalidations) {
+        if (result.resolvedPath != null) {
+          graph.invalidateOnFileUpdate(subrequestNode.id, result.resolvedPath);
+        }
+
+        for (let filePath of result.includedFiles) {
+          graph.invalidateOnFileUpdate(subrequestNode.id, filePath);
+        }
+
+        if (result.watchGlob != null) {
+          graph.invalidateOnFileCreate(subrequestNode.id, result.watchGlob);
+        }
+
+        if (result.shouldInvalidateOnStartup) {
+          graph.invalidateOnStartup(subrequestNode.id);
+        }
+      }
+      subrequestNodes.push(subrequestNode);
+
+      // Add dep version requests
+      for (let [moduleSpecifier, version] of result.devDeps) {
+        let depVersionRequst = {
+          moduleSpecifier,
+          resolveFrom: result.resolvedPath, // TODO: resolveFrom should be nearest package boundary
+        };
+        let id = generateRequestId('dep_version_request', depVersionRequst);
+        let shouldSetupInvalidations =
```

Comment:
nit: inline or move closer to its use below

Reply:
https://github.com/parcel-bundler/parcel/pull/4345#discussion_r395880413